target_chains/solana/programs/pyth-price-publisher/src/accounts/buffer.rs

```diff
@@ -0,0 +1,209 @@
+    pub confidence:                    u64,
+}
+
+impl BufferedPrice {
+    pub fn new(
+        feed_index: u32,
+        trading_status: u32,
+        price: i64,
+        confidence: u64,
+    ) -> Result<Self, PublisherPriceError> {
+        if feed_index == 0 || feed_index >= (1 << 28) || trading_status >= (1 << 4) {
+            return Err(PublisherPriceError);
+        }
+        Ok(Self {
+            trading_status_and_feed_index: (trading_status << 28) | feed_index,
+            price,
+            confidence,
+        })
+    }
+
+    pub fn trading_status(&self) -> u32 {
+        self.trading_status_and_feed_index >> 28
+    }
+
+    pub fn feed_index(&self) -> u32 {
+        self.trading_status_and_feed_index & ((1 << 28) - 1)
+    }
+}
+
+pub fn format_matches(data: &[u8]) -> bool {
+    super::format(data).map_or(false, |f| f == FORMAT)
+}
+
+pub fn read(data: &[u8]) -> Result<(&BufferHeader, &[BufferedPrice]), ReadAccountError> {
+    if data.len() < size_of::<BufferHeader>() {
+        return Err(ReadAccountError::DataTooShort);
+    }
+    let header: &BufferHeader = from_bytes(&data[..size_of::<BufferHeader>()]);
+    if header.format != FORMAT {
+        return Err(ReadAccountError::FormatMismatch);
+    }
+    let prices_bytes = &data[size_of::<BufferHeader>()..];
+    let num_prices: usize = header.num_prices.try_into().unwrap();
+    let expected_len = num_prices.saturating_mul(size_of::<BufferedPrice>());
+    if expected_len > prices_bytes.len() {
+        return Err(ReadAccountError::InvalidNumPrices);
+    }
+    // We don't validate the values of `new_prices` to make the publishing process
+    // more efficient. They will be validated when applied in the validator.
+    let prices = cast_slice(&prices_bytes[..expected_len]);
+    Ok((header, prices))
+}
+
+pub fn size(max_prices: usize) -> usize {
+    size_of::<BufferHeader>() + max_prices * size_of::<BufferedPrice>()
+}
+
+pub fn read_mut(data: &mut [u8]) -> Result<(&mut BufferHeader, &mut [u8]), ReadAccountError> {
+    if data.len() < size_of::<BufferHeader>() {
+        return Err(ReadAccountError::DataTooShort);
+    }
+    let (header, prices) = data.split_at_mut(size_of::<BufferHeader>());
+    let header: &mut BufferHeader = from_bytes_mut(header);
+    if header.format != FORMAT {
+        return Err(ReadAccountError::FormatMismatch);
+    }
+    Ok((header, prices))
+}
+
+pub fn create(
+    data: &mut [u8],
+    publisher: [u8; 32],
+) -> Result<(&mut BufferHeader, &mut [u8]), ReadAccountError> {
+    if data.len() < size_of::<BufferHeader>() {
+        return Err(ReadAccountError::DataTooShort);
+    }
+    let (header, prices) = data.split_at_mut(size_of::<BufferHeader>());
+    let header: &mut BufferHeader = from_bytes_mut(header);
+    if header.format != 0 {
+        return Err(ReadAccountError::AlreadyInitialized);
+    }
+    *header = BufferHeader::new(publisher);
+    Ok((header, prices))
+}
+
+pub fn extend(
```

Comment:
sort of a nitpick but I think "extend" isn't a good name for this -- "extend" usually implies that you will add more data without overwriting existing data. I think "update" or "update_prices" would be a better name.